### ISSUE TITLE
Several games that Failed to compile shaders

### ISSUE THREAD

**Issue by JohnGodgames** [RPCS3/rpcs3#2328]
# Current progress

- [x] Fixed games
- [ ] Not fixed games yet / not retested

**GAMES LIST**
- [ ] RailFan

Already fixed games by commit https://github.com/RPCS3/rpcs3/pull/2567

- [x] Kamen Rider Battride War
- [x] Kamen Rider Battride War II
- [x] Iron man 2
- [x] Tales of Xillia
- [x] Wheelman

Already fixed games by commit https://github.com/RPCS3/rpcs3/pull/2482
- [x] Full Auto 2: Battlelines
- [x] Kane & Lynch: Dead Men
- [x] Lost Dimension
- [x] Midnight Club Los Angeles
- [x] Metal Gear Solid HD Collection
- [x] MotoGP08
- [x] RESIDENT EVIL REVELATIONS 2 TRIAL
- [x] SBK08
- [x] SBK09

Already fixed by time:
- [x] Asura's Wrath
- [x] Tales of Xillia 2

**Full Auto 2: Battlelines**

```E {rsx::thread} RSX: Failed to compile shader: 0(70) : error C7011: implicit cast from "float" to "vec4"
F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
-------------
0(70) : error C7011: implicit cast from "float" to "vec4"
(0) : error C2003: incompatible options for link
```

[Shader log](https://mega.nz/#!mMEjxK7Y!CBzIxkkRDsg72TAzwiWJwnMCg8BJyx6niAqYljkbric)
[Log](https://mega.nz/#!vdMDUKjI!PebFZQZwo2cYijtslXy5jOkwDlcPrOmSPl3E-hAYT1Y)

**Midnight Club Los Angeles**

```
E {rsx::thread} RSX: Failed to compile shader: 0(261) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec3, float, float)"
0(262) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec3, float, float)"

F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
-------------
0(261) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec3, float, float)"
0(262) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec3, float, float)"
(0) : error C2003: incompatible options for link
```

[Shaderlog](https://mega.nz/#!vckWXB5D!j181ZRYBygZyAsxKwa6qH5is6SgauPSbojTEbI9wbXE)
[Log](https://mega.nz/#!GAFnSTga!W3b-IZ0IWAZBlkxl2IbFxPY_sk2ik5HVYc-F9CTrhUw)

**MotoGP08 - (BLUS30220)**
```
·F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
-------------
0(174) : error C0000: syntax error, unexpected '.', expecting "::" at token "."
0(175) : error C0000: syntax error, unexpected '.', expecting "::" at token "."
(0) : error C2003: incompatible options for link
```
[LOG](https://mega.nz/#!SIs3hYjI!GG2veN6N2THEOEeHzJHc3zYYlnLO4TuQLevr4lwYlek)

**SBK08 - (BLES00368)**
```
F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
-------------
0(304) : error C0000: syntax error, unexpected '.', expecting ')' at token "."
0(306) : error C0000: syntax error, unexpected '.', expecting "::" at token "."
(0) : error C2003: incompatible options for link
```
[LOG](https://mega.nz/#!7VMQhJSA!C_OEGiXA1Kq5-QJ3eFIggSneHhLf-ZqdHVBi43PJclg)

**SBK09**

```
E {rsx::thread} RSX: Failed to compile shader: 0(76) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec3, float, float)"
0(77) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec3, float, float)"
0(81) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec3, float, float)"

F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
-------------
0(76) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec3, float, float)"
0(77) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec3, float, float)"
0(81) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec3, float, float)"
(0) : error C2003: incompatible options for link
```

[Shaderlog](https://mega.nz/#!mVlECTTa!6WcgzOzTJzAq9zhVOp7ijaRNM-mxxfa3VPP2FYxumeU)
[Log](https://mega.nz/#!KAc2CbTQ!aXtdp6VsRU3wxQoGO-9oyZbwWK1KQhcRxJBSYKUkfxM)

**Comment by aakh1361** [comment]
RESIDENT EVIL REVELATIONS 2 TRIAL -(NPEB02128)

> F  {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
0(129) : error C1115: unable to find compatible overloaded function "textureLod(sampler1D, float, vec4)"
(0) : error C2003: incompatible options for link
'
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/801537/RPCS3.zip)

**Comment by aakh1361** [comment]
Metal Gear Solid HD Collection
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/804609/RPCS3.zip)

> E {rsx::thread} RSX: Failed to compile shader: 0(73) : error C1008: undefined variable "r1"
F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
0(73) : error C1008: undefined variable "r1"
(0) : error C2003: incompatible options for link

**Comment by aakh1361** [comment]
RESIDENT EVIL REVELATIONS 2 TRIAL [NPEB02128]
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/817332/RPCS3.zip)

> E {rsx::thread} RSX: Failed to compile shader: 0(196) : error C1115: unable to find compatible overloaded function "textureLod(sampler1D, float, vec4)"

> F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
0(196) : error C1115: unable to find compatible overloaded function "textureLod(sampler1D, float, vec4)"
(0) : error C2003: incompatible options for link

**Comment by aakh1361** [comment]
Kane & Lynch: Dead Men [NPEB90045]

[RRPCS3.zip](https://github.com/RPCS3/rpcs3/files/818194/RRPCS3.zip)

> E {rsx::thread} RSX: Failed to compile shader: 0(63) : error C7011: implicit cast from "float" to "vec4"
> F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
0(63) : error C7011: implicit cast from "float" to "vec4"
(0) : error C2003: incompatible options for link

**Comment by kd-11** [comment]
Some of these are really just copy-paste bugs. Will be fixed ASAP

**Comment by kd-11** [comment]
Please retest with https://github.com/RPCS3/rpcs3/pull/2482

**Comment by Zangetsu38** [comment]
No change for Midnight
>E {rsx::thread} RSX: Warning, version 420 is not yet complete; most version-specific features are present, but some are missing.
ERROR: 0:256: 'textureGrad' : no matching overloaded function found 
ERROR: 0:256: 'assign' :  cannot convert from 'const float' to 'inout 4-component vector of float'
ERROR: 0:257: 'textureGrad' : no matching overloaded function found 
ERROR: 3 compilation errors.  No code generated.
E {rsx::thread} RSX: 
F {rsx::thread} class std::runtime_error thrown: Failed to compile fragment shader
(in file Emu\RSX\VK\VKFragmentProgram.cpp:461)

**Comment by kd-11** [comment]
@Zangetsu attach the fragment shader that caused the crash, or test with opengl and attach rpcs3.log

**Comment by Zangetsu38** [comment]
ok
[Midnight VK.zip](https://github.com/RPCS3/rpcs3/files/824252/Midnight.VK.zip)

**Comment by kd-11** [comment]
Thanks. Looks like textureGrad needs more tinkering

**Comment by Zangetsu38** [comment]
In OGL
>E {rsx::thread} RSX: Failed to compile shader: 0(261) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec2, float, float)"
0(262) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec2, float, float)"

F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
-------------
0(261) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec2, float, float)"
0(262) : error C1115: unable to find compatible overloaded function "textureGrad(sampler2D, vec2, float, float)"
(0) : error C2003: incompatible options for link
[midnight OGL.zip](https://github.com/RPCS3/rpcs3/files/824264/midnight.OGL.zip)

**Comment by kd-11** [comment]
should be fixed now with latest commit.

**Comment by Zangetsu38** [comment]
Ok testing :)

**Comment by Zangetsu38** [comment]
Works in Vulkan have menu and gamne launch with video, but very slow, i have a big  regression speed before just for first logo 2 fps, before 30 

![capture d ecran 2017-03-07 13 08 02](https://cloud.githubusercontent.com/assets/5261759/23655861/4dd9fd9e-0337-11e7-8e09-338060720144.png)
![capture d ecran 2017-03-07 13 02 13](https://cloud.githubusercontent.com/assets/5261759/23655860/4dd7418a-0337-11e7-91b3-f0738fba9efd.png)

**Comment by kd-11** [comment]
There should not be a performance difference since the changes are limited to unimplemented shader instructions.

**Comment by Zangetsu38** [comment]
i say before for several month ^^, i have recorded video for this, and intro speed normal, but now very slow

**Comment by kd-11** [comment]
ahh. ok. Its probably the new changes to Cell emulation.

**Comment by digitaldude555** [comment]
![screenshot 911](https://cloud.githubusercontent.com/assets/26138653/23656867/1750aefa-0344-11e7-963f-b514372da6db.png)

Got rid of that error on Full Auto 2 Battlelines, now it goes to intro instead of nothing.

**Comment by digitaldude555** [comment]
```
F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
-------------
0(86) : error C1031: swizzle mask element not present in operand "z"
(0) : error C2003: incompatible options for link

```

I get this for motogp08 when going ingame.
[Log](https://www.mediafire.com/?v6hn8jjndh1m181)

**Comment by SakataGintokiYT** [comment]
@kd-11 
[e189d146]

**LOST DIMENSION**


**OpenGL**
```
E {rsx::thread} RSX: Failed to compile shader: 0(119) : error C1115: unable to find compatible overloaded function "texture(samplerCube, vec2, float)"

F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
-------------
0(119) : error C1115: unable to find compatible overloaded function "texture(samplerCube, vec2, float)"
(0) : error C2003: incompatible options for link
'
```
**[Log:](https://gist.github.com/SakataGintokiYT/39f83cd65ae616d691d1137fce01f32a)**


**Vulkan**
```
E {rsx::thread} RSX: Warning, version 420 is not yet complete; most version-specific features are present, but some are missing.
ERROR: 0:117: 'texture' : no matching overloaded function found 
ERROR: 0:117: 'xyz' : vector field selection out of range 
ERROR: 0:117: 'assign' :  cannot convert from 'const float' to 'temp 3-component vector of float'
ERROR: 3 compilation errors.  No code generated.


E {rsx::thread} RSX: 
F {rsx::thread} class std::runtime_error thrown: Failed to compile fragment shader
(in file C:\rpcs3\rpcs3\Emu\RSX\VK\VKFragmentProgram.cpp:461)

```
**[Log:](https://gist.github.com/SakataGintokiYT/37b088cc32bc297f2849e4a7749f5acb)**

**Comment by digitaldude555** [comment]
![screenshot 922](https://cloud.githubusercontent.com/assets/26138653/23664966/d2dc0de8-035f-11e7-9241-419ab8bbbb75.png)
![screenshot 923](https://cloud.githubusercontent.com/assets/26138653/23664970/d6393164-035f-11e7-9ad3-780632276289.png)

kane and lynch dead men goes ingame now.

**Comment by kd-11** [comment]
Any games on this list still not working? I know that there will be some "src type 3 used" errors, but that is a separate issue altogether.

**Comment by kd-11** [comment]
motogp08 should be fixed with latest commit. Working on the lost dimension one.

**Comment by digitaldude555** [comment]
![screenshot 930](https://cloud.githubusercontent.com/assets/26138653/23668340/6e675f2a-0369-11e7-9d09-748e98c0c522.png)

Yes motogp08 is fixed now, it goes ingame.

**Comment by kd-11** [comment]
Lost dimension should be fixed too

**Comment by digitaldude555** [comment]
![screenshot 937](https://cloud.githubusercontent.com/assets/26138653/23669581/5eaf58b8-036d-11e7-86da-ef7f4c4dc69b.png)

Lost dimension goes ingame now. It's fixed.

**Comment by kd-11** [comment]
I think only MGS is left on the list. Not sure if I can fix it before we can merge the PR.

**Comment by JohnGodgames** [comment]
**Tales of Xillia**

```
E {rsx::thread} RSX: Src type 3 used, please report this to a developer.
E {rsx::thread} RSX: Src type 3 used, please report this to a developer.
E {rsx::thread} RSX: Src type 3 used, please report this to a developer.
E {rsx::thread} RSX: Src type 3 used, please report this to a developer.
E {rsx::thread} RSX: Failed to compile shader: 0(71) : error C0000: syntax error, unexpected '.', expecting ')' at token "."

F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
-------------
0(71) : error C0000: syntax error, unexpected '.', expecting ')' at token "."
(0) : error C2003: incompatible options for link
```

[LOG](https://mega.nz/#!vVkmSKKZ!6okjeRxe1yOzmQuFa59HkJlX_-oTE4dBK2nSVNMNpvo)

**Comment by digitaldude555** [comment]
![screenshot 946](https://cloud.githubusercontent.com/assets/26138653/23670390/3b750cfa-0370-11e7-9dbc-43d7b9933a26.png)

Resident Evil Revelations 2 shows the opening movie and is black ingame, the error is gone though and it hasn't reached this far before.

**Comment by kd-11** [comment]
Sorry, can't fix src type 3 so quickly. Needs reverse engineering the instructions, although I have some ideas what it might be.

**Comment by JohnGodgames** [comment]
so we can make a separate issue for leftovers, but also need to check for regressions before merge.I'm testing a several working games now.

**Comment by Xcedf** [comment]
MGS HD works to the menu, but it's black screen with sound, i've tested this game long ago, it was showing few intros but graphic output for this game was also broken long ago

**Comment by aakh1361** [comment]
@kd-11
I wedding tomorrow and don't have time now to test games
If you don't have Hurry for merge 
I can 2-3 days test very more games in my library for test shader error

**Comment by JohnGodgames** [comment]
Looks like a regression

**Iron man 2**

```
E {rsx::thread} RSX: Failed to compile shader: 0(63) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(64) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(65) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(66) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(68) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(69) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(70) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(71) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"

F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
-------------
0(63) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(64) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(65) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(66) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(68) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(69) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(70) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
0(71) : error C0000: syntax error, unexpected reserved word "break", expecting "::" at token "break"
(0) : error C2003: incompatible options for link
```

[LOG](https://mega.nz/#!DF83QKzJ!0GUEEBY09GwxCrOzNfRZvETcPqZ3Nm20jAacWyJgPl0)

**Comment by Zangetsu38** [comment]
Nier Fixed, but have glitch texture ^^ 
https://youtu.be/rz6i26JbE94

**Comment by Zangetsu38** [comment]
RailFan
E {rsx::thread} RSX: Unknown/illegal instruction: 0x6a (forced unit 0)
E {rsx::thread} RSX: Src type 3 used, please report this to a developer.
E {rsx::thread} RSX: Src type 3 used, please report this to a developer.
E {rsx::thread} RSX: Warning, version 420 is not yet complete; most version-specific features are present, but some are missing.
ERROR: 0:64: '' :  syntax error
ERROR: 1 compilation errors.  No code generated.
E {rsx::thread} RSX: 
F {rsx::thread} class std::runtime_error thrown: Failed to compile fragment shader
(in file Emu\RSX\VK\VKFragmentProgram.cpp:467)
[RainFan.zip](https://github.com/RPCS3/rpcs3/files/826635/RainFan.zip)

OGL
>E {rsx::thread} RSX: Unknown/illegal instruction: 0x6a (forced unit 0)
E {rsx::thread} RSX: Src type 3 used, please report this to a developer.
E {rsx::thread} RSX: Src type 3 used, please report this to a developer.
E {rsx::thread} RSX: Failed to compile shader: 0(63) : error C0000: syntax error, unexpected '.', expecting ')' at token "."
F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
0(63) : error C0000: syntax error, unexpected '.', expecting ')' at token "."
(0) : error C2003: incompatible options for link

**Comment by raven02** [comment]
Probably consolidate Vulkan one that fail to compile shader for After Burner Mix https://github.com/RPCS3/rpcs3/issues/2407

**Comment by Zangetsu38** [comment]
MS : PR DX12
@kd-11 Tested with you last commit.
[MS PR DX12.zip](https://github.com/RPCS3/rpcs3/files/839263/MS.PR.DX12.zip)
>E {rsx::thread} RSX: FS build failed:D:\Emulateurs\Rpcs3\FragmentProgram.hlsl(132,10-19): error X3014: incorrect number of arguments to numeric-type constructor
F {rsx::thread} class std::runtime_error thrown: fragment program compilation failure
(in file g:\git\rpcs3-dante\rpcs3\emu\rsx\d3d12\D3D12PipelineState.h:162)

**Comment by Zangetsu38** [comment]
Last PR railfan not fixed, Nier broken with color rose and several game broken :/

**Comment by kd-11** [comment]
Those are all src type 3 errors. This is a known bug and is tracked in another issue.
Its not difficult to "work around" the compilation issue, and I will do so as soon as I'm done fixing regressions caused by last commit.

**Comment by Zangetsu38** [comment]
Asura's Wrath freeze after few sec ingame with this
E {rsx::thread} RSX: Src type 3 used, please report this to a developer.
E {rsx::thread} RSX: Warning, version 420 is not yet complete; most version-specific features are present, but some are missing.
ERROR: 0:67: '' :  syntax error
ERROR: 1 compilation errors.  No code generated.
E {rsx::thread} RSX: 
F {rsx::thread} class std::runtime_error thrown: Failed to compile fragment shader
(in file Emu\RSX\VK\VKFragmentProgram.cpp:472)

**Comment by SakataGintokiYT** [comment]
@kd-11 
[2b79a369]
**KamenRider Battlide War**

**OpenGL:**
```
E {rsx::thread} RSX: Failed to compile shader: 0(64) : error C1008: undefined variable "tex0"
0(64) : error C1008: undefined variable "tc0"
0(64) : error C1008: undefined variable "tex0_coord_scale"

F {rsx::thread} class gl::glsl::link_exception thrown: linkage failed: 'Fragment info
-------------
0(64) : error C1008: undefined variable "tex0"
0(64) : error C1008: undefined variable "tc0"
0(64) : error C1008: undefined variable "tex0_coord_scale"
(0) : error C2003: incompatible options for link
'
```
Log:

**Vulkan:**
```
E {rsx::thread} RSX: Warning, version 420 is not yet complete; most version-specific features are present, but some are missing.
ERROR: 0:89: 'tex0' : undeclared identifier 
ERROR: 0:89: 'tc0' : undeclared identifier 
ERROR: 0:89: 'xy' : vector field selection out of range 
ERROR: 0:89: 'texture' : no matching overloaded function found 
ERROR: 0:89: 'a' : vector field selection out of range 
ERROR: 5 compilation errors.  No code generated.


E {rsx::thread} RSX: 
F {rsx::thread} class std::runtime_error thrown: Failed to compile fragment shader
(in file C:\rpcs3\rpcs3\Emu\RSX\VK\VKFragmentProgram.cpp:472)
```
Log:

**Comment by kd-11** [comment]
Some instances should be fixed with https://github.com/RPCS3/rpcs3/pull/2803

**Comment by scribam** [comment]
@JohnGodgames Would you mind ticking the checkbox for the following games? They seem to work now.
- Asura's Wrath (https://www.youtube.com/watch?v=wXX7TV8RMaw)
- Tales of Xillia 2 (https://www.youtube.com/watch?v=4XTnJuD5NX4)

Anyone knows if RailFan still have the issue?

**Comment by TotalCaesar659** [comment]
I've wanted to test Railfan and got white screen only on both OpenGL and Vulkan.
[Railfan_(OpenGL).txt](https://github.com/RPCS3/rpcs3/files/1220062/Railfan_.OpenGL.txt)
[RailFan_(Vulkan).txt](https://github.com/RPCS3/rpcs3/files/1220063/RailFan_.Vulkan.txt)

**Comment by jjsat** [comment]
Deus Ex Human Revolution Directors Cut (BLES01928) cannot go ingame with Vulkan due to this (happens while switching from the loading screen to ingame):
```
E {rsx::thread} RSX: ERROR: 0:156: 'texture' : no matching overloaded function found 
ERROR: 0:156: '' : compilation terminated 
ERROR: 2 compilation errors.  No code generated.


E {rsx::thread} RSX: 
F {rsx::thread} class std::runtime_error thrown: Failed to compile fragment shader
(in file Emu\RSX\VK\VKFragmentProgram.cpp:466)
```

Works fine on OpenGL.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1360667/RPCS3.log.gz)

**Comment by AniLeo** [comment]
We need the last game "RailFan" to be retested.

**Comment by Megamouse** [comment]
railfan is menu now with `F {PPU[0x1000088] Thread (HLE Video Decoder)} class std::runtime_error thrown: AU decoding error(0xbebbb1b7)
(in file Emu\Cell\Modules\cellVdec.cpp:243)`

**Comment by AniLeo** [comment]
Closing as solved

**Comment by RenanZX** [comment]
Dragon Ball raging blast(Androids Saga after the first battle) using Vulkan
```
E {rsx::thread} RSX: 
F {rsx::thread} class std::runtime_error thrown: Failed to compile fragment shader
(in file Emu\RSX\VK\VKFragmentProgram.cpp:467)
```

**Comment by Carlosss18** [comment]
Initial D Extreme Stage 
E {RSX Decompiler Thread} RSX: Failed to compile shader: ERROR: 0:121: 'tex00' : undeclared identifier 
ERROR: 0:121: 'texture' : no matching overloaded function found (using implicit conversion) 
ERROR: 0:121: 'process_texel' : no matching overloaded function found (using implicit conversion) 
ERROR: 0:121: 'assign' :  cannot convert from 'const highp float' to 'inout 4-component vector of highp float'
